src/discovery/mdns.js

```diff
@@ -0,0 +1,119 @@
+import { TypedEmitter } from 'tiny-typed-emitter'
+import net from 'node:net'
+import NoiseSecretStream from '@hyperswarm/secret-stream'
+import { once } from 'node:events'
+import dnssd from 'dnssd'
+import b4a from 'b4a'
+
+
+const SERVICE_NAME = 'mapeo'
+
+
+/** @typedef {{ publicKey: Buffer, secretKey: Buffer }} Keypair */
+
+/**
+ * @typedef {Object} Service
+ * @property {string} host - hostname of the service
+ * @property {number} port - port of the service
+ * @property {Object<string, any>} txt - TXT records of the service
+ * @property {string[]} [addresses] - addresses of the service
+ */
+
+
+export class MdnsDiscovery extends TypedEmitter {
+  #identityKeypair
+  #server
+  /** @type {Set<string>} */
+  #connections = new Set()
+  /** @type {typeof import('dnssd').Advertisement} */
+  #advertiser
+  /** @type {typeof import('dnssd').Browser} */
+  #browser
+
+  /** @param {Object} opts
+   * @param {Keypair} opts.identityKeypair
+   * @param {Number} opts.port */
+  constructor({ identityKeypair, port }) {
+    super()
+    this.#identityKeypair = identityKeypair
+    this.#server = net.createServer(this.#handleConnection.bind(this, false))
+  }
+
+  async start() {
+    // Let OS choose port, listen on ip4, all interfaces
+    this.#server.listen(0, '0.0.0.0')
+    // TODO: listen for errors
+    await once(this.#server, 'listening')
+
+    const addr = /** @type {net.AddressInfo} */ (this.#server.address())
+    const listeningPort = typeof addr === 'string' ? addr : addr?.port
+
+
+    this.#advertiser = new dnssd.Advertisement(
+      dnssd.tcp(SERVICE_NAME),
+      listeningPort
+    )
+    await this.#advertiser.start()
+
+    // find all peers adverticing Mapeo
+    this.#browser = await dnssd
+      .Browser(dnssd.tcp(SERVICE_NAME))
+      .on('serviceUp',
+        /** @param {Service} service */
+        (service) => {
+          const socket = net.connect(
+            service.port,
+            service.addresses && service.addresses[0]
+          )
+          socket.on('connect',() => {
+            this.#handleConnection(true, socket)
+          })
+        })
+      .start()
+  }
+
+  /**
+   * @param {boolean} isInitiator
+   * @param {net.Socket} socket
+   */
+  async #handleConnection(isInitiator, socket) {
+    if(!socket.remoteAddress) return
+
+    const remoteAddress = socket.remoteAddress
+    if (this.#connections.has(remoteAddress)) {
+      socket.destroy()
+      return
+    }
+
+    const secretStream = new NoiseSecretStream(isInitiator, socket, {
+      keyPair: this.#identityKeypair,
+    })
+
+    secretStream.on('close', () => {
+      this.#connections.delete(remoteAddress)
+    })
+
+
+    secretStream.on('connect', async () => {
+      // deterministically choose the connection to keep,
+      // by ordering the {remote,local}public keys
+      const keep = !secretStream.isInitiator ||
+        b4a.compare(secretStream.publicKey, secretStream.remotePublicKey) > 0
+      if(keep){
```

Comment:
We don't want to choose to keep based on `isInitiator` - both sides of the connection will make the same determination.

The reason for this code Hyperswarm is to check for duplicate outgoing connections, e.g. a device A tries to connect to device B twice. This can happen if device B starts advertising, so device A connects, then device B stops advertising then starts again, without closing the original connection. Then device A will see the advertising re-appear and re-try to connect.

To solve this I think we actually need to change `this.#noiseConnections` to a `Map<remotePublicKey, noiseStream>`. Then within `secretStream.on('connect')` we can check:

1. is there a previous connection where we were the initiator?
2. if we were the initiator of this new connection, then drop the new connection (e.g. keep the existing one)